src/main/java/dev/blaauwendraad/masker/json/MaskingState.java

```diff
@@ -250,19 +293,86 @@ private void checkCurrentValueBounds(int index) {
         }
     }
 
+    /**
+     * Flushes the current buffer into provided OutputStream, then moves the current token to the beginning of the buffer
+     * and fills up the buffer from provided InputStream.
+     * In case the current token is too long (the start index is not in the last quarter of the buffer),
+     * extends the current buffer by a quarter.
+     *
+     * @return true if more data is available in the input stream, false if the end of the stream is reached.
+     */
+    boolean readNextBuffer() {
+        if (inputStream == null || outputStream == null) {
+            return false;
+        }
+
+        // flush the remaining of the current buffer up to the current value start index into output stream
+        try {
+            int end = currentValueStartIndex == -1 ? messageLength : currentValueStartIndex;
+            outputStream.write(message, lastReplacementEndIndex, end - lastReplacementEndIndex);
+            outputStream.flush();
+            lastReplacementEndIndex = 0;
+        } catch (IOException e) {
+            throw new UncheckedIOException(STREAM_READ_ERROR_MESSAGE, e);
+        }
+
+        int currentValueLength = currentValueStartIndex == -1 ? 0 : messageLength - currentValueStartIndex;
+        if (currentValueLength == 0) {
+            // the pointer is not at a json value, so we are safe to read the next buffer
+            currentIndex -= messageLength;
+            try {
+                messageLength = inputStream.readNBytes(message, 0, bufferSize);
+            } catch (IOException e) {
+                throw new UncheckedIOException(STREAM_READ_ERROR_MESSAGE, e);
+            }
+        } else {
+            // the current buffer has ended before the masker finished processing the current value.
+
+            if (currentValueLength < bufferSize >> 2) {
+                // in case the current value is shorter than a quarter of the buffer,
+                // fill up the buffer without extending its length
+
+                // move the current value to the beginning of the buffer
+                System.arraycopy(message, currentValueStartIndex, message, 0, currentValueLength);
+            } else {
+                // in case the current value is longer than a quarter of the buffer,
+                // extend the buffer length by a quarter
+                bufferSize += bufferSize >> 2;
+                byte[] extension = new byte[bufferSize];
+
+                // move the current value to the beginning of the extension
+                System.arraycopy(message, currentValueStartIndex, extension, 0, currentValueLength);
+                message = extension;
```

Comment:
I'd rather extend by x2, instead of a quarter. While this is highly unlikely, it's generally how most collections amortize the complexity.

Another question: what happens if extending the buffer once isn't enough?

Reply:
Good idea

> Another question: what happens if extending the buffer once isn't enough?

The current value length is always less than the size of the current buffer. So we keep increasing current buffer size every iteration untill we reach the end of the value.

Reply:
Done